enterprise/server/crypter_service/crypter_service.go

```diff
@@ -0,0 +1,291 @@
+	return []byte(fmt.Sprintf("%24s", groupID))
+}
+
+func makeChunkAuthHeader(chunkIndex uint32, groupIDHeader []byte) ([]byte, error) {
+	if 4+len(groupIDHeader) != encryptedChunkAuthHeaderSize {
+		return nil, status.FailedPreconditionError("invalid header size")
+	}
+
+	chunkHeader := make([]byte, encryptedChunkAuthHeaderSize)
+	binary.LittleEndian.PutUint32(chunkHeader, chunkIndex)
+	copy(chunkHeader[4:], groupIDHeader)
+	return chunkHeader, nil
+}
+
+func (e *Encryptor) flushBlock() error {
+	if e.bufIdx == 0 {
+		return nil
+	}
+
+	if _, err := rand.Read(e.nonceBuf); err != nil {
+		return err
+	}
+	if _, err := e.w.Write(e.nonceBuf); err != nil {
+		return err
+	}
+
+	e.chunkCounter++
+	chunkAuth, err := makeChunkAuthHeader(e.chunkCounter, e.groupIDHeader)
+	if err != nil {
+		return err
+	}
+	ct := e.ciph.Seal(e.buf[:0], e.nonceBuf, e.buf[:e.bufIdx], chunkAuth)
+	if _, err := e.w.Write(ct); err != nil {
+		return err
+	}
+	e.bufIdx = 0
+	return nil
+}
+
+func (e *Encryptor) Write(p []byte) (n int, err error) {
+	if !e.wroteHeader {
+		if _, err := e.w.Write([]byte(encryptedDataHeaderSignature)); err != nil {
+			return 0, err
+		}
+		if _, err := e.w.Write([]byte{encryptedDataHeaderVersion}); err != nil {
+			return 0, err
+		}
+		e.wroteHeader = true
+	}
+
+	readIdx := 0
+	for readIdx < len(p) {
+		readLen := e.bufCap - e.bufIdx
+		if readLen > len(p)-readIdx {
+			readLen = len(p) - readIdx
+		}
+		copy(e.buf[e.bufIdx:], p[readIdx:readIdx+readLen])
+		e.bufIdx += readLen
+		readIdx += readLen
+		if e.bufIdx == e.bufCap {
+			if err := e.flushBlock(); err != nil {
+				return 0, err
+			}
+		}
+	}
+
+	return len(p), nil
+}
+
+func (e *Encryptor) Commit() error {
+	if err := e.flushBlock(); err != nil {
+		return err
+	}
+	return e.w.Commit()
+}
+
+func (e *Encryptor) Close() error {
+	return e.w.Close()
+}
+
+type Decryptor struct {
+	ciph            cipher.AEAD
+	groupIDHeader   []byte
+	r               io.ReadCloser
+	headerValidated bool
+	chunkCounter    uint32
+
+	// buf contains the decrypted plaintext ready to be read.
+	buf []byte
+	// bufIdx is the index at which the plaintext can be read.
+	bufIdx int
+	// bufLen is the amount of plaintext in the buf ready to be read.
+	bufLen int
+}
+
+func (d *Decryptor) Read(p []byte) (n int, err error) {
+	if !d.headerValidated {
+		fileHeader := make([]byte, 3)
+		if _, err := d.r.Read(fileHeader); err != nil {
+			return 0, err
+		}
+		if string(fileHeader[0:2]) != encryptedDataHeaderSignature {
+			return 0, status.InternalErrorf("invalid file signature %d %d", fileHeader[0], fileHeader[1])
+		}
+		if fileHeader[2] != encryptedDataHeaderVersion {
+			return 0, status.InternalErrorf("invalid file version %d", fileHeader[2])
+		}
+		d.headerValidated = true
+	}
+
+	// No plaintext available, need to decrypt another chunk.
+	if d.bufIdx >= d.bufLen {
+		n, err := io.ReadFull(d.r, d.buf)
+		// We expect a short read on the last chunk.
+		if err != nil && err != io.ErrUnexpectedEOF {
```

Comment:
Instead of the short-read error check, it might be slightly clearer to use a regular `Read` and then allow `io.EOF` only if the value of `n` plus the number of bytes we've read so far matches the full stream length. That way it doesn't look like we're ignoring an unexpected error